src/System.Text.ValueBuilder/System/Variant.cs

```diff
@@ -0,0 +1,444 @@
+        /// <summary>
+        /// Get the value as an object if the value is stored as an object.
+        /// </summary>
+        /// <param name="value">The value, if an object, or null.</param>
+        /// <returns>True if the value is actually an object.</returns>
+        public bool TryGetValue(out object value)
+        {
+            bool isObject = Type == VariantType.Object;
+            value = isObject ? _object : null;
+            return isObject;
+        }
+
+        /// <summary>
+        /// Get the value as the requested type <typeparamref name="T"/> if actually stored as that type.
+        /// </summary>
+        /// <param name="value">The value if stored as (T), or default.</param>
+        /// <returns>True if the <see cref="Variant"/> is of the requested type.</returns>
+        public unsafe bool TryGetValue<T>(out T value) where T : unmanaged
+        {
+            value = default;
+            bool success = false;
+
+            // Checking the type gets all of the non-relevant compares elided by the JIT
+            if((typeof(T) == typeof(bool) && Type == VariantType.Boolean)
+                || (typeof(T) == typeof(byte) && Type == VariantType.Byte)
+                || (typeof(T) == typeof(char) && Type == VariantType.Char)
+                || (typeof(T) == typeof(DateTime) && Type == VariantType.DateTime)
+                || (typeof(T) == typeof(DateTimeOffset) && Type == VariantType.DateTimeOffset)
+                || (typeof(T) == typeof(decimal) && Type == VariantType.Decimal)
+                || (typeof(T) == typeof(double) && Type == VariantType.Double)
+                || (typeof(T) == typeof(Guid) && Type == VariantType.Guid)
+                || (typeof(T) == typeof(short) && Type == VariantType.Int16)
+                || (typeof(T) == typeof(int) && Type == VariantType.Int32)
+                || (typeof(T) == typeof(long) && Type == VariantType.Int64)
+                || (typeof(T) == typeof(sbyte) && Type == VariantType.SByte)
+                || (typeof(T) == typeof(float) && Type == VariantType.Single)
+                || (typeof(T) == typeof(TimeSpan) && Type == VariantType.TimeSpan)
+                || (typeof(T) == typeof(ushort) && Type == VariantType.UInt16)
+                || (typeof(T) == typeof(uint) && Type == VariantType.UInt32)
+                || (typeof(T) == typeof(ulong) && Type == VariantType.UInt64))
+            {
+                // The JIT is able to generate more efficient code when including the
+                // code for CastTo<T>() directly.
+                fixed (void* u = &_union)
+                {
+                    value = *(T*)u;
+                }
+                success = true;
+            }
+
+            return success;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private unsafe T CastTo<T>() where T : unmanaged
+        {
+            fixed (void* u = &_union)
+            {
+                return *(T*)u;
+            }
```

Comment:
Is it better than `Unsafe.As<TFrom, TTo>`?

Reply:
Forgot this field is readonly so I can't easily get a ref to it. I'd be contorting so much it just seemed wiser to use fixed.

Reply:
It is what `Unsafe.AsRef` is for.

Reply:
> It is what Unsafe.AsRef is for.

Yes, I just thought it was a little too ugly, but not hung up on it.  The interesting thing, however, is that `Unsafe.As<Union, T>(ref Unsafe.AsRef(_union));` is much more efficient, which surprises me (@AndyAyersMS). I'm not terribly great at this level so perhaps it shouldn't.

```
00007ffa`a48133b0 System.Variant.CastTo[[System.Int32, System.Private.CoreLib]]()
00007ffa`a48133b7 3909            cmp     dword ptr [rcx],ecx
00007ffa`a48133b9 4883c110        add     rcx,10h
00007ffa`a48133bd 48890c24        mov     qword ptr [rsp],rcx
00007ffa`a48133c1 488b0424        mov     rax,qword ptr [rsp]
00007ffa`a48133c5 8b00            mov     eax,dword ptr [rax]
```

Goes to:

```
00007ffa`a48234a0 System.Variant.CastTo[[System.Int32, System.Private.CoreLib]]()
00007ffa`a48234a0 488d4110        lea     rax,[rcx+10h]
00007ffa`a48234a4 8b00            mov     eax,dword ptr [rax]
```

Also doesn't suffer from the inlining problem the fixed does. Changing...